workload/schemachange: `ALTER FUNCTION` may fail to find functions without arguments

As of https://github.com/cockroachdb/cockroach/pull/115359, the RSW supports `ALTER FUNCTION`.

On occasion, `ALTER FUNCTION` statements will fail:
```
test logs left over in: /home/chrisseto/go/src/github.com/cockroachdb/cockroach/tmp/_tmp/e2bfcb2daef366fa77f42af32e17a96d/logTestWorkload2821957365
--- FAIL: TestWorkload (35.85s)
    test_log_scope.go:170: test logs captured to: /home/chrisseto/go/src/github.com/cockroachdb/cockroach/tmp/_tmp/e2bfcb2daef366fa77f42af32e17a96d/logTestWorkload2821957365
    test_log_scope.go:81: use -show-logs to present logs inline
    schema_change_external_test.go:132:
                Error Trace:    pkg/ccl/testccl/workload/schemachange/schemachange_test_test/pkg/ccl/testccl/workload/schemachange/schema_change_external_test.go:132
                Error:          Received unexpected error:
                                ***UNEXPECTED ERROR; Received an unexpected execution error.: ERROR: unknown function: public.udf_84() (SQLSTATE 42883)
                                (1) forced error mark
                                  | "fatal error when running txn"
                                  | github.com/cockroachdb/errors/withstack/*withstack.withStack::
                                Wraps: (2)
                                Wraps: (3) attached stack trace
                                  -- stack trace:
                                  | github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*opStmt).executeStmt
                                  |     github.com/cockroachdb/cockroach/pkg/workload/schemachange/operation_generator.go:3087
                                  | github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*schemaChangeWorker).runInTxn
                                  |     github.com/cockroachdb/cockroach/pkg/workload/schemachange/schemachange.go:446
                                  | github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*schemaChangeWorker).run
                                  |     github.com/cockroachdb/cockroach/pkg/workload/schemachange/schemachange.go:523
                                  | pkg/ccl/testccl/workload/schemachange/schemachange_test_test.TestWorkload.TestWorkload.func4.func6
                                  |     pkg/ccl/testccl/workload/schemachange/schemachange_test_test/pkg/ccl/testccl/workload/schemachange/schema_change_external_test.go:121
                                  | golang.org/x/sync/errgroup.(*Group).Go.func1
                                  |     golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:75
                                  | runtime.goexit
                                  |     src/runtime/asm_amd64.s:1650
                                Wraps: (4) ***UNEXPECTED ERROR; Received an unexpected execution error.
                                Wraps: (5) ERROR: unknown function: public.udf_84() (SQLSTATE 42883)
                                Error types: (1) *markers.withMark (2) *schemachange.ErrorState (3) *withstack.withStack (4) *errutil.withPrefix (5) *pgconn.PgError
                Test:           TestWorkload
    schema_change_external_test.go:98: backup, tracing data, and system table dumps in /home/chrisseto/go/src/github.com/cockroachdb/cockroach/tmp/_tmp/e2bfcb2daef366fa77f42af32e17a96d/logTestWorkload2821957365
    panic.go:523: -- test log scope end --
```

Unfortunately, my debug info got GC'd. I'll upload any that I manage to reproduce. From the runs that I've inspected, the referenced function exists with no arguments in the `public` schema. Notably, this issue only occurs in the `public` schema. This may be a red herring if `ALTER FUNCTION SET SCHEMA` never succeeds. 



Jira issue: CRDB-34777

Epic CRDB-19168

--- a/pkg/workload/schemachange/generate.go
+++ b/pkg/workload/schemachange/generate.go
@@ -187,6 +187,10 @@ func Generate[T tree.Statement](
 		aIsSuccess := a.Code == pgcode.SuccessfulCompletion
 		bIsSuccess := b.Code == pgcode.SuccessfulCompletion
 
+		if aIsSuccess && bIsSuccess {
+			// If both are valid, choose randomly.
+			return rng.Float64() < 0.5
+		}
 		return aIsSuccess && !bIsSuccess
 	})
 

--- a/pkg/workload/schemachange/operation_generator.go
+++ b/pkg/workload/schemachange/operation_generator.go
@@ -198,7 +198,7 @@ func (og *operationGenerator) randOp(
 			// We can only ignore this error, if no other PgErrors
 			// were set in the clean up process.
 			if errors.Is(err, pgx.ErrNoRows) &&
-				!errors.Is(err, &pgconn.PgError{}) {
+				!errors.HasType(err, &pgconn.PgError{}) {
 				continue
 			}
 			// Table select had a primary key swap, so no statement
@@ -3878,8 +3878,15 @@ func (og *operationGenerator) createFunction(ctx context.Context, tx pgx.Tx) (*o
 				COALESCE(member->>'direction' = 'REMOVE', false) AS dropping
 			FROM enum_members
 		`)
+	schemasQuery := With([]CTE{
+		{"descriptors", descJSONQuery},
+	}, `SELECT quote_ident(name) FROM descriptors WHERE descriptor ? 'schema'`)
 
 	enums, err := Collect(ctx, og, tx, pgx.RowToMap, enumQuery)
+	if err != nil {
+		return nil, err
+	}
+	schemas, err := Collect(ctx, og, tx, pgx.RowTo[string], schemasQuery)
 	if err != nil {
 		return nil, err
 	}
@@ -3916,6 +3923,9 @@ func (og *operationGenerator) createFunction(ctx context.Context, tx pgx.Tx) (*o
 			name := tree.Name(fmt.Sprintf("udf_%s", og.newUniqueSeqNumSuffix()))
 			return &name
 		},
+		"Schema": func() (string, error) {
+			return PickOne(og.params.rng, schemas)
+		},
 		"DroppingEnum": func() (string, error) {
 			return PickOne(og.params.rng, droppingEnums)
 		},
@@ -3947,9 +3957,9 @@ func (og *operationGenerator) createFunction(ctx context.Context, tx pgx.Tx) (*o
 	// to the schema workload but may become a nice to have.
 	stmt, expectedCode, err := Generate[*tree.CreateRoutine](og.params.rng, og.produceError(), []GenerationCase{
 		// 1. Nothing special, fully self contained function.
-		{pgcode.SuccessfulCompletion, `CREATE FUNCTION { UniqueName } (i int, j int) RETURNS VOID LANGUAGE SQL AS $$ SELECT NULL $$`},
+		{pgcode.SuccessfulCompletion, `CREATE FUNCTION { Schema } . { UniqueName } (i int, j int) RETURNS VOID LANGUAGE SQL AS $$ SELECT NULL $$`},
 		// 2. 1 or more table or type references spread across parameters, return types, or the function body.
-		{pgcode.SuccessfulCompletion, `CREATE FUNCTION { UniqueName } ({ ParamRefs }) RETURNS { ReturnRefs } LANGUAGE SQL AS $$ SELECT NULL WHERE { BodyRefs } $$`},
+		{pgcode.SuccessfulCompletion, `CREATE FUNCTION { Schema } . { UniqueName } ({ ParamRefs }) RETURNS { ReturnRefs } LANGUAGE SQL AS $$ SELECT NULL WHERE { BodyRefs } $$`},
 		// 3. Reference a table that does not exist.
 		{pgcode.UndefinedTable, `CREATE FUNCTION { UniqueName } () RETURNS VOID LANGUAGE SQL AS $$ SELECT * FROM "ThisTableDoesNotExist" $$`},
 		// 4. Reference a UDT that does not exist.
@@ -4092,7 +4102,19 @@ func (og *operationGenerator) alterFunctionSetSchema(
 	functionsQuery := With([]CTE{
 		{"descriptors", descJSONQuery},
 		{"functions", functionDescsQuery},
-	}, `SELECT quote_ident(schema_id::REGNAMESPACE::TEXT) || '.' || quote_ident(name) FROM functions`)
+	}, `SELECT
+			quote_ident(schema_id::REGNAMESPACE::TEXT) || '.' || quote_ident(name) || '(' || array_to_string(funcargs, ', ') || ')'
+			FROM functions
+			JOIN LATERAL (
+				SELECT
+					COALESCE(array_agg(quote_ident(typnamespace::REGNAMESPACE::TEXT) || '.' || quote_ident(typname)), '{}') AS funcargs
+				FROM pg_catalog.pg_type
+				JOIN LATERAL (
+					SELECT unnest(proargtypes) AS oid FROM pg_catalog.pg_proc WHERE oid = (id + 100000)
+				) args ON args.oid = pg_type.oid
+			) funcargs ON TRUE
+			`,
+	)
 
 	schemasQuery := With([]CTE{
 		{"descriptors", descJSONQuery},

--- a/pkg/workload/schemachange/optype.go
+++ b/pkg/workload/schemachange/optype.go
@@ -262,10 +262,10 @@ var opWeights = []int{
 	alterDatabaseDropSuperRegion:      0, // Disabled and tracked with #111299
 	alterDatabasePrimaryRegion:        0, // Disabled and tracked with #83831
 	alterDatabaseSurvivalGoal:         0, // Disabled and tracked with #83831
-	alterFunctionRename:               0, // Disabled and tracked with #116794.
-	alterFunctionSetSchema:            0, // Disabled and tracked with #116794.
+	alterFunctionRename:               1,
+	alterFunctionSetSchema:            1,
 	alterTableAddColumn:               1,
-	alterTableAddConstraintForeignKey: 1, // Tentatively re-enabled, see #91195.
+	alterTableAddConstraintForeignKey: 1,
 	alterTableAddConstraintUnique:     0,
 	alterTableAlterColumnType:         0, // Disabled and tracked with #66662.
 	alterTableAlterPrimaryKey:         1,
@@ -288,7 +288,7 @@ var opWeights = []int{
 	createTableAs:                     1,
 	createTypeEnum:                    1,
 	createView:                        1,
-	dropFunction:                      0, // Disabled and tracked with #116794.
+	dropFunction:                      1,
 	dropIndex:                         1,
 	dropSchema:                        0, // Disabled and tracked with 116792.
 	dropSequence:                      1,

